core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/DriverComponent.scala

```diff
@@ -119,6 +120,48 @@ trait DriverComponent {
   def renameForHiding(recordCount: Long, name: String): String = {
     name + "_" + getSufficientlyRandomSuffix(recordCount)
   }
+
+  /**
+    * Converts a `Seq[(A, B)]` into a `Map[A, B]`, combining the values with a `Monoid[B]` in case of key conflicts.
+    *
+    * For example:
+    * {{{
+    * scala> groupPairs(Seq(("a", 1), ("b", 2), ("a", 3)))
+    * res0: Map[String,Int] = Map(b -> 2, a -> 4)
+    * }}}
+    * */
+  def groupPairs[A, B: Monoid](pairs: Seq[(A, B)]): Map[A, B] =
+    pairs.toList.foldMap { case (a, b) => Map(a -> b) }
+
+  /**
+    * Converts a `Seq[(A, F[B])]` into a `Map[A, F[B]]`, combining the values with the _universal_ monoid for type F.
+    * This can be useful for when using groupBy with slick aggregate functions such as max, min, avg, etc which return
+    * an Option.
+    *
+    * For example:
+    * {{{
+    * scala> case class Foo(i: Int)
+    * defined class Foo
+    *
+    * scala> groupPairs(Seq(("a", Foo(1).some), ("b", Foo(2).some), ("c", Foo(3).some)))
+    * << does not compile as there is no Monoid instance for Foo >>
+    *
+    * scala> groupPairsK(Seq(("a", Foo(1).some), ("b", Foo(2).some), ("c", Foo(3).some)))
+    * res9: Map[String,Option[Foo]] = Map(b -> Some(Foo(2)), a -> Some(Foo(1)), c -> Some(Foo(3)))
+    * }}}
+    */
+  def groupPairsK[F[_], A, B](pairs: Seq[(A, F[B])])(implicit M: MonoidK[F]): Map[A, F[B]] =
+    groupPairs(pairs)(M.algebra[B])
```

Comment:
Thanks for the feedback here. After reflecting a bit, maybe it would have been better to leave `groupByWorkspaceIdThenStatus` and `groupByWorkspaceId` as is -- I didn't really need to change them for this bug. I'm still learning how to scala in a large team, and readability is important.

Digging in to your points:

0. +1
1. see below
2. ok
3. Huh, I haven't seen that before. I had to enable the `-feature` compiler flag to see that warning. I guess it's complaining about the `F[_]` type parameter. The [scaladoc](http://www.scala-lang.org/api/2.10.3/#scala.language$) is kind of funny:
> Why control it? Higher kinded types in Scala lead to a Turing-complete type system, where compiler termination is no longer guaranteed. They tend to be useful mostly for type-level computation and for highly generic design patterns. The level of abstraction implied by these design patterns is often a barrier to understanding for newcomers to a Scala codebase. Some syntactic aspects of higher-kinded types are hard to understand for the uninitiated and type inference is less effective for them than for normal types. Because we are not completely happy with them yet, it is possible that some aspects of higher-kinded types will change in future versions of Scala. So an explicit enabling also serves as a warning that code involving higher-kinded types might have to be slightly revised in the future.

4,5. So here's why I added the K version. In this case I wanted to group triples where the key value is an `Option[java.sql.Timestamp]`. Therefore I need a Monoid instance for `java.sql.Timestamp`, which doesn't exist. It wouldn't make much sense to define a Monoid for timestamps -- what would you do, add them together? However, in this case I'm guaranteed to have no key conflicts because of the SQL structure (`group by`, etc). So I can just use a Monoid for Option which just takes one or the other, regardless of the value inside the "box". That's exactly what `MonoidK[Option]` does.

Furthermore, I'm not sure how I would implement this in terms of `groupTriples` without using `MonoidK`. Perhaps this would be more clear?
```
private def groupByWorkspaceIdThenStatus(workflowDates: Seq[(UUID, String, Option[Timestamp])]): Map[UUID, Map[String, Option[Timestamp]]] = {
  // bla bla comment about bringing a monoid into scope with Option.orElse behavior
  implicit val optionUniversalMonoid: Monoid[Option[Timestamp]] = MonoidK[Option].algebra[Timestamp]
  CollectionUtils.groupTriples(workflowDates)
}
```
Then at least the `MonoidK` stuff is localized in a private method with some explanation specific to the use case.